tortoise/state.go

```diff
@@ -511,3 +508,26 @@ func activeSetWeight(epoch *epochInfo, aset []types.ATXID) (uint64, error) {
 	}
 	return weight, nil
 }
+
+type layerSlice struct {
+	data []*layerInfo
+}
+
+func (s *layerSlice) get(offset, index types.LayerID) *layerInfo {
+	i := index - offset - 1
+	lth := types.LayerID(len(s.data))
+	if i < lth {
+		return s.data[i]
+	}
+	last := offset + lth
+	for j := types.LayerID(0); j <= i-lth; j++ {
+		s.data = append(s.data, &layerInfo{lid: last + j + 1, opinions: map[types.Hash32]votes{}})
+	}
+	return s.get(offset, index)
+}
+
+func (s *layerSlice) pop() {
+	copy(s.data[:], s.data[1:])
```

Comment:
how about container.List? remove(front()) without the copying?

Reply:
linked list doesn't support constant cost o(1) reads